### PR TITLE
xl: Fix rare freeze after many disk/network errors

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
xl.storageDisks is passed to some low-level XL functions where some disks of
xl.storageDisks are set to nil when they encounter some errors. This means all
elements in xl.storageDisks will be nil after some time which lead to an unusable XL.

## Motivation and Context
Manual testing.

I manually modified posix.go to return randomly errors before I discovered this bug.

## How Has This Been Tested?
Manual testing

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/xl-v1-utils.go [hunk 1]
```diff
@@ -357,6 +357,23 @@ func shuffleDisks(disks []StorageAPI, distribution []int) (shuffledDisks []Stora
 	return shuffledDisks
 }
 
+// evalDisks - returns a new slice of disks where nil is set if
+// the correspond error in errs slice is not nil
+func evalDisks(disks []StorageAPI, errs []error) []StorageAPI {
+	if len(errs) != len(disks) {
```
Comment (harshavardhana) [inline cmd/xl-v1-utils.go:363]: This should never happen, returning a nil should accompany a log message here with errorIf() indicating something really unexpected has happened.

File: cmd/xl-v1-utils.go [hunk 2]
```diff
@@ -357,6 +357,23 @@ func shuffleDisks(disks []StorageAPI, distribution []int) (shuffledDisks []Stora
...
+		return nil
```
Comment (harshavardhana) [inline cmd/xl-v1-utils.go]: Add a test case for this as well separately with its own unit tests..

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by vadmeste** [comment]
Comments adressed

File: cmd/xl-v1-utils_test.go [hunk 3]
```diff
@@ -429,3 +430,53 @@ func testShuffleDisks(t *testing.T, xl *xlObjects) {
...
+
+	diskErr := errors.New("some disk error")
+	errs := []error{
+		diskErr, nil, nil, nil,
+		nil, diskErr, nil, nil,
+		diskErr, nil, nil, nil,
+		nil, nil, nil, diskErr,
+	}
+
+	newDisks := evalDisks(disks, errs)
```
Comment (harshavardhana) [inline cmd/xl-v1-utils_test.go]: why not have a test for errs to be smaller than disks?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by vadmeste** [comment]
Fixed, can we also add more reviewers ? @deekoder

**Review by harshavardhana — DISMISSED** [review]

**Review by krishnasrinivas — COMMENTED** [review]
@vadmeste xl.storageDisks is never supposed to get modified because listOnlineDisks() and shuffleDisks() always create new slice. Need to review where exactly xl.storageDisks get modified, can you check?

**Comment by vadmeste** [comment]
> @vadmeste xl.storageDisks is never supposed to get modified because listOnlineDisks() and shuffleDisks() always create new slice. Need to review where exactly xl.storageDisks get modified, can you check?

https://github.com/minio/minio/pull/4438/commits/2b5e0ac17eecc8c825487e59ed16775e3684432d#diff-949516a155a3e0e700699d2ff77d6fb9L504

This is one example.

I still think the new style is better than the old one. Modifying a slice deep inside many function levels is not good as the developer will always forget about this behavior.

**Comment by krishnasrinivas** [comment]
```
577                 partSizeWritten, checkSums, erasureErr := erasureCreateFile(onlineDisks, minioMetaTmpBucket, tempErasureObj, partReader\
    , allowEmptyPart, partsMetadata[0].Erasure.BlockSize, partsMetadata[0].Erasure.DataBlocks, partsMetadata[0].Erasure.ParityBlocks, bitRo\
    tAlgo, xl.writeQuorum)
```

Since this does not return `[]StorageAPI` we might end up with missing parts because later we do:
```
		if onlineDisks, err = renameXLMetadata(onlineDisks, minioMetaTmpBucket, tempObj, srcBucket, srcObject, xl.writeQuorum); err != nil {
  			return ObjectInfo{}, toObjectErr(err, srcBucket, srcObject)		  			return ObjectInfo{}, toObjectErr(err, srcBucket, srcObject)
  		}		  		}
```

I will be reviewing more.

**Comment by krishnasrinivas** [comment]
We can make sure that we follow this pattern everywhere:

```
function() {                                                                                                                                                                                                       
        onlineDisks, err := f(onlineDisks, ....)                                                                                                                                                                   
        onlineDisks, err := g(onlineDisks, ....)                                                                                                                                                                   
        onlineDisks, err := h(onlineDisks, ....)                                                                                                                                                                   
}
```

There are two exceptions, one in PutObject and the other in CompleteMultipartUpload where it uses xl.storageDisks to purge the previously uploaded object (before renaming from tmp location to the actual location)

Here are the list of comments:

* erasureCreateFile should return `disks`
* In PutObject and CompleteMultipartUpload, after erasureCreateFile() subsequent calls should use the `disks` returned by erasureCreateFile
* In newMultipartUpload() you are ignoring returned disks from writeSameXLMetadata() and renameObject() There is no harm in this particular situation. But to be consistent you need not ignore, you can use the disks returned by writeSameXLMetadata() in renameObject() to be consistent. Also it will be more performant as renameObject() will not try on disks that are down.

**Comment by vadmeste** [comment]
All your comments are addressed @krishnasrinivas

**Review by krishnasrinivas — APPROVED** [review]

**Comment by krisis** [comment]
@vadmeste could you provide the steps to manually test this PR? Also, what is the behaviour user observes when more than quorum number of `xl.StorageDisks` are `nil` in current master (in your manual testing)?

**Comment by vadmeste** [comment]
> could you provide the steps to manually test this PR? Also, what is the behaviour user observes when more than quorum number of xl.storageDisks are nil in current master (in your manual testing)?

I'll post the patch in some minutes. I simply added some lines in posix.go which returns an error with a chance of 10% in some calls (rename, delete, append, ..)

**Comment by vadmeste** [comment]
[occasional_disk_error.patch.txt](https://github.com/minio/minio/files/1052063/occasional_disk_error.patch.txt)

This is the patch (remove .txt) which returns fake disk errors.

To reproduce:
```
$ cd $GOPATH/src/github.com/minio/minio
$ patch -p1 < occasional_disk_error.patch
$ go install
$ minio server /tmp/xl{1..4}
```
In another terminal
```
$ mc mb myminio/testbucket
$ dd if=/dev/urandom of=6mb.file bs=1024 count=$[1024*6]
$ while true; do mc cp 6mb.file myminio/testbucket ; done
```
After some time, you'll notice that mc will never be able to upload again.

By the way, I've just noticed that I put freeze word to describe the issue, it is not 100% correct, I meant  unusable instead.

**Review by krisis — APPROVED** [review]
Tested locally with @vadmeste patch in comments.

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**